tez-mapreduce/src/main/java/org/apache/hadoop/mapred/split/TezGroupedSplitsInputFormat.java

```diff
@@ -129,14 +138,50 @@ void initInputFormatFromSplit(TezGroupedSplit split) throws TezException {
     int idx = 0;
     long progress;
     RecordReader<K, V> curReader;
-    
+    AtomicInteger initIndex;
+    int numReaders;
+    ExecutorService initReaderExecService;
+    Queue<Future<RecordReader<K,V>>> initedReaders;
+
     public TezGroupedSplitsRecordReader(TezGroupedSplit split, JobConf job,
         Reporter reporter) throws IOException {
       this.groupedSplit = split;
       this.job = job;
       this.reporter = reporter;
+      this.initIndex = new AtomicInteger(0);
+      int numThreads = conf.getInt(TezSplitGrouper.TEZ_GROUPING_SPLIT_INIT_THREADS,
+          TezSplitGrouper.TEZ_GROUPING_SPLIT_INIT_THREADS_DEFAULT);
+      this.numReaders = conf.getInt(TezSplitGrouper.TEZ_GROUPING_SPLIT_INIT_NUM_RECORDREADERS,
+          TezSplitGrouper.TEZ_GROUPING_SPLIT_INIT_NUM_RECORDREADERS_DEFAULT);
+      this.initReaderExecService = Executors.newFixedThreadPool(numThreads,
```

Comment:
I think we will want to somehow make this static (to share between usages of TezGroupedSplitsRecordReader) or figure out how to call shutdown in a reliable manner. Otherwise I think a long lived process that uses multiple TezGroupedSplitsRecordReaders throughout its life will end up having a large number of unused threads since they will not auto shutdown and thus not garbage collected.

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html
See finalization section.

Reply:
I'm not sure about the context, but if the goal is to properly shutdown a global resource somehow, it can done by the shutdownhandler, somewhere here: https://github.com/apache/tez/blob/master/tez-dag/src/main/java/org/apache/tez/dag/app/DAGAppMaster.java#L954